---
title: "10-UNIPROTproteomics"
author: "Kathryn Walters"
date: "2023-07-19"
output: html_document
---

```{r libraries, warning=FALSE, message=FALSE, echo=FALSE}
library(Biostrings)
library(rlist)
library(ggVennDiagram)
library(readr)
library(here)
library(dplyr)
library(tibble)
library(tidyr)
library(ggplot2)
library(ggthemes)
library(EBSeq)
# library(conflicted)
# conflict_prefer("here", "here")
# conflict_prefer("filter", "dplyr")
```


```{r load data, message=FALSE, warning=FALSE, echo=FALSE}

geneInfo <- read_csv(here("accessories","gencode.v26.primary.info.csv.zip"), col_names = F, show_col_types = FALSE) 
colnames(geneInfo) <- c("gene_id","transcript_id","biotype","symbol")

tx2gene <- geneInfo[,c(2,1)]
```


```{r proteomics load, warning=FALSE, message=FALSE, echo=FALSE}
proteomicsCyto2 <- read_tsv(here("proteomics", "UNIPROTalignment", "Cyto24_set2_protein.tsv")) %>% as.data.frame()
proteomicsCyto2 <- proteomicsCyto2[,c(6,15,20, 26:35)]
proteomicsCyto2 <- proteomicsCyto2 %>% filter(`Unique Peptides` > 1 & `Total Intensity` > 0)  %>% na.omit()

proteomicsCyto3 <- read_tsv(here("proteomics", "UNIPROTalignment", "Cyto24_set3_protein.tsv")) %>% as.data.frame()
proteomicsCyto3 <- proteomicsCyto3[,c(6,15,20, 26:35)]
proteomicsCyto3 <- proteomicsCyto3 %>% filter(`Unique Peptides` > 1 & `Total Intensity` > 0)  %>% na.omit()

proteomicsEndoC <- read_tsv(here("proteomics", "UNIPROTalignment", "EndoC_IFNa24_protein.tsv")) %>% as.data.frame()
proteomicsEndoC <- proteomicsEndoC[,c(6,15,20, 26:35)]
proteomicsEndoC <- proteomicsEndoC %>% filter(`Unique Peptides` > 1 & `Total Intensity` > 0) %>% na.omit()

proteomicsSBCs <- read_tsv(here("proteomics", "UNIPROTalignment", "sBC_LFQ_Uniprotmatch_protein.tsv")) %>% as.data.frame()
proteomicsSBCs <- proteomicsSBCs[,c(4,14,26,27)]
proteomicsSBCs$Intensity <- proteomicsSBCs$`nORF1 Total Intensity` + proteomicsSBCs$`nORF2 Total Intensity`

proteomicsSBCs <- proteomicsSBCs %>% filter(`Combined Unique Spectral Count` > 1 & Intensity > 0) %>% na.omit()


```

```{r venndiagram, message=FALSE, warning=FALSE}
proteomicsIslets_genes <- rbind(proteomicsCyto2[1], proteomicsCyto3[1]) %>% as.data.frame() %>% unique()


proteomicsSBCs_genes <- proteomicsSBCs[1] %>% as.data.frame() %>% unique()

proteomicsEndoC_genes <- proteomicsEndoC[1] %>% as.data.frame() %>% unique()

set1 <- as.matrix(proteomicsIslets_genes)[,1]
set2 <- as.matrix(proteomicsSBCs_genes)[,1]
set3 <- as.matrix(proteomicsEndoC_genes)[,1]

listALL <- list(EndoC = set3, Islets = set1, sBCs = set2)


p_proteomCellLineoverlap <- ggVennDiagram(listALL, edge_color = "black", label = "count", label_alpha = 0) + theme(legend.position = "none")

p_proteomCellLineoverlap

#ggsave(plot = p_proteomCellLineoverlap, filename = here("plots", "proteomicsCellLines_overlap_venndiagram.pdf"), width = 4, height = 3)

intersection <- intersect(intersect(set1, set2), set3)

intersection <- as.data.frame(intersection)
colnames(intersection) <- "Gene"

intersection <- left_join(intersection, proteomicsCyto2)
intersection <- intersection[,-c(2,3)]

intersection <- left_join(intersection, proteomicsCyto3)
intersection <- intersection[,-c(12,13)] %>% na.omit()

intersection <- left_join(intersection, proteomicsEndoC)
intersection <- intersection[,-c(22,23)] %>% na.omit()

intersection <- left_join(intersection, proteomicsSBCs)
intersection <- intersection[,-c(32,35)] %>% na.omit()

intersection <- intersection %>%
  dplyr::group_by(intersection$Gene) %>%
  summarise(across(.cols = -Gene, .fns = sum))

test <- as.data.frame(intersection)

intersection <- test %>% column_to_rownames("intersection$Gene")
```


```{r}
intersection <- intersect(intersect(set1, set2), set3)

intersection <- as.data.frame(intersection)
colnames(intersection) <- "Gene"

intersection <- left_join(intersection, proteomicsCyto2)
intersection <- intersection[,c(1,3)]

intersection <- left_join(intersection, proteomicsCyto3, by = join_by(Gene))
intersection <- intersection[,c(1,2,4)] %>% na.omit()

intersection <- left_join(intersection, proteomicsEndoC)
intersection <- intersection[,c(1,2,3,5)] %>% na.omit()

intersection <- left_join(intersection, proteomicsSBCs, by = join_by(Gene))
intersection <- intersection[,c(1,2,3,4,8)] %>% na.omit()

intersection <- intersection %>%
  dplyr::group_by(intersection$Gene) %>%
  summarise(across(.cols = -Gene, .fns = mean))

test <- as.data.frame(intersection)

intersection <- test %>% column_to_rownames("intersection$Gene")
```



## Clustering samples


```{r clustering qc, echo=FALSE, message=FALSE, warning=FALSE}
# use mature gene level (sum of all mature transcripts) estimates for QC EDA
intersection[intersection==0] <- NA
intersection <- intersection %>% na.omit()

test <- MedianNorm(intersection)

write_csv(intersection, here("output", "heatmapCorrDataset.csv"))

qcinput <- log2(intersection)

qcCor <- cor(qcinput, method = 'spearman')

rownames(qcCor) <- colnames(qcCor)

#below we are generating a heatmap to show that each replicate clusters together and that FshB KO are different than the WT and Het FshB KO. It saves directly to file rather than showing in R. 
nCut <- 2 
pheatmap <- pheatmap::pheatmap(qcCor, clustering_distance_rows = "euclidean",clustering_distance_cols = "euclidean", clustering_method = "complete", cutree_cols = nCut, cutree_rows = nCut, border_color = "black", filename = here("plots", "CellLine_heatmap.pdf"))


Fpheatmap::pheatmap(qcCor, clustering_distance_rows = "euclidean",clustering_distance_cols = "euclidean", clustering_method = "complete", cutree_cols = nCut, cutree_rows = nCut, border_color = "black")


ggsave()

```


## PCA analysis


```{r pca qc, echo=FALSE, message=FALSE, warning=FALSE}
pca_data <- prcomp(qcinput, center = T, scale. = T) # perform pca

pca_data_info <- summary(pca_data) # summarize the PCs by variance and save object

pca_plot_data <- data.frame(pca_data$rotation) # we make a dataframe out of the rotations and will use this to plot

pca_plot_data$ID <- rownames(pca_plot_data)

pca_plot_data <- pca_plot_data %>% separate(col = ID, sep = "_", into = c("geno","rep"))



pca_plot_data$geno <- relevel(as.factor(pca_plot_data$geno), ref = "Cyto")


labelPosition <- pca_plot_data %>% group_by(geno, rep) %>% dplyr::select(PC1, PC2) %>% summarise(mPC1=mean(PC1), mPC2=mean(PC2))

myCols <- c("darkgrey","darkred", "blue", "purple")



p_pca <- ggplot(pca_plot_data, aes(x=PC1, y=PC2, color=geno)) +
  geom_point(size=2) + 
  theme_minimal() +
  xlab(paste("PC1 (%",100*round(pca_data_info$importance[2,1], digits = 3),")", sep = "")) +
  ylab(paste("PC2 (%",100*round(pca_data_info$importance[2,2], digits = 3),")", sep = "")) +
  scale_color_manual(values = myCols)

p_pca

ggsave(plot = p_pca, filename = here("plots", "allCells_proteo_PCAplot.pdf"), device = "pdf", units = "in", width = 6, height = 4, dpi = 320)
```




