---
title: "12-nanoPotsProteomics"
author: "Kathryn Walters"
date: "2023-07-20"
output: html_document
---


```{r load data, message=FALSE, warning=FALSE, echo=FALSE}
AllORFs <- read_csv(here("output", "GeneCollapsedORFs_AllPrograms.csv"))

geneInfo <- read_csv(here("accessories","gencode.v26.primary.info.csv.zip"), col_names = F, show_col_types = FALSE) 
colnames(geneInfo) <- c("gene_id","transcript_id","biotype","symbol")

tx2gene <- geneInfo[,c(2,1)]
```



```{r message=FALSE, warning=FALSE, echo=FALSE}
#getting RiboCode list of ORFs found in proteomics
proteomicsRC_nano <- read_tsv(here("proteomics", "nanoPots_data", "RiboCode", "combined_protein.tsv")) %>% filter(`Combined Unique Spectral Count` > 1) %>% pull(`Protein ID`) %>% as.data.frame()
colnames(proteomicsRC_nano) <- "ORF_ID"


#getting Ribotricer list of ORFs found in proteomics
proteomicsRT_nano <- read_tsv(here("proteomics", "nanoPots_data", "Ribotricer", "combined_protein.tsv")) %>% filter(`Combined Unique Spectral Count` > 1) %>% pull(`Protein ID`) %>% as.data.frame()
colnames(proteomicsRT_nano) <- "ORF_ID"


#getting ORFquant list of ORFs found in proteomics
proteomicsOQ_nano <- read_tsv(here("proteomics", "nanoPots_data", "ORFQuant", "combined_protein.tsv")) %>% filter(`Combined Unique Spectral Count` > 1) %>% pull(`Protein ID`) %>% as.data.frame()
colnames(proteomicsOQ_nano) <- "ORF_ID"

```


```{r tidy proteomics, message=FALSE, warning=FALSE, echo=FALSE}
#Need to switch the ORF ID from transcripts to genes for Ribotricer and ORFquant
 proteomicsRT <- proteomicsRT %>% separate(col = "ORF_ID", into = c("transcript_id", "first", "second", "third"), sep = "_", remove = TRUE) 
 proteomicsRT <- left_join(proteomicsRT, tx2gene) %>% unite(col = "ORF_ID", c("gene_id", "first", "second", "third"), sep = "_", remove = TRUE) %>% select("ORF_ID")
 

 proteomicsOQ <- proteomicsOQ %>% separate(col = "ORF_ID", into = c("transcript_id", "first", "second"), sep = "_", remove = TRUE) 
 proteomicsOQ <- left_join(proteomicsOQ, tx2gene) %>% unite(col = "ORF_ID", c("gene_id", "first", "second"), sep = "_", remove = TRUE) %>% select("ORF_ID")

```



```{r message=FALSE, warning=FALSE, echo=FALSE}

```

