---
title: "14a-GettingNAseqs"
author: "Kathryn Walters"
date: "2023-08-01"
output: html_document
---


```{r libraries, warning=FALSE, message=FALSE, echo=FALSE}
library(Biostrings)
library(rlist)
library(tximport)
library(ggVennDiagram)
library(readr)
library(here)
library(dplyr)
library(tidyr)
library(tibble)
library(ggplot2)
library(ggthemes)
library(reshape2)
library(readxl)
library(BSgenome.Hsapiens.UCSC.hg38)
library(seqinr)
library(stringr)
# library(conflicted)
# 
# conflict_prefer("here", "here")
# conflict_prefer("filter", "dplyr")
```


```{r message=FALSE, warning=FALSE}
#grabbing the metadata to make txi file.
metadata <- readxl::read_xlsx(here::here("accessories","NMLabLibrarySummary.xlsx"), skip = 1) %>%
  filter(Project == "Holger" & Index_1_i7!="PooledLibraries")

#keeping only the beta cell data
metadata <- metadata[18:23,]

#simplifing the information
metadata <- data.frame(
  sample=metadata$SampleID, genotype = metadata$Genotype,
  type=factor(case_when(metadata$LibraryPrepKit=="Qiagen_miRNA" ~ "ribo",
          T ~ "rna")),
  rep=factor(case_when(metadata$Treatment2 == "RepA" ~ "A",
                       T ~ "B")
             )
)

metadata[1:2,2] <- "GFP+"
#making sure the name is correct and will be able to access folder properly.
metadata$sample <- case_when(metadata$type=="ribo" ~ paste(metadata$sample,"_R2",sep = ""),
          T ~ metadata$sample)

myquantfiles1 <- paste("data/",
                      metadata$sample,
                      "/quant.sf",
                      sep = "")

myquantfiles2 <- paste("data/",
                      metadata$sample,
                      "/quant.sf",
                      sep = "")

myquantfiles <- append(myquantfiles1[1:2], myquantfiles2[3:6])
  
names(myquantfiles) <- paste(metadata$genotype, metadata$type, metadata$rep, sep = "_")
```


```{r load data, message=FALSE, warning=FALSE, echo=FALSE}
AllORFs <- read_xlsx(here("output", "AllORFs_wProteomics.xlsx"))

geneInfo <- read_csv(here("accessories","gencode.v26.primary.info.csv.zip"), col_names = F, show_col_types = FALSE) 
colnames(geneInfo) <- c("gene_id","transcript_id","biotype","symbol")

tx2gene <- geneInfo[,c(2,1)]

```


```{r functions}
sequences <- readDNAStringSet(here("accessories", "gencode.v26.transcripts.fa.gz"))

df <- data.frame(
  Name = names(sequences),
  Sequence = as.character(sequences),
  stringsAsFactors = FALSE
)

# Split the column based on "|"
df <- separate(df, Name, into = paste0("element", 1:8), sep = "\\|", extra = "drop")


getNAseq <- function(transcriptID, start_position, end_position) {

specific_row_text <- df %>%
  filter(element1 == transcriptID) %>%
  dplyr::slice(1) %>%
  mutate(extracted_text = substr(Sequence, start_position, end_position)) %>%
  pull(extracted_text)


# Return the result
return(specific_row_text)
}


getKozakseq <- function(transcriptID, start_position) {

tStart <- start_position - 6
tEnd <- start_position +3

specific_row_text <- df %>%
  filter(element1 == transcriptID) %>%
  dplyr::slice(1) %>%
  mutate(extracted_text = substr(Sequence, tStart, tEnd)) %>%
  pull(extracted_text)


# Return the result
return(specific_row_text)
}
```

```{r}
#Optimal codons were chosen to be the codons with maximum occurrences for each amino acid in highly expressed genes (20% of the genes with the highest PA). -Got that from this paper:
#https://www.sciencedirect.com/science/article/pii/S2001037021001549#b0195
#Since this is part of the codon bias calculation, I need to get amino acid sequences for these as well. 

#So essentially I need to find this. 
#Find out the top 20% of genes with biggest TE. 
TE_TPM <- read_xlsx(here("output", "TPM_TEdata.xlsx"))
TE_TPM <- TE_TPM %>% unique()

TE_TPM <- TE_TPM %>% arrange(desc(TE))
TE_TPM <- TE_TPM %>% filter(biotype == "protein_coding")

p20TE <- TE_TPM[1:round(nrow(TE_TPM)*.2),] %>% pull(gene_id)

#The problem so far is that these are at the gene level. However I need the transcript level. So I'm going to grab the most abundant transcript. 

#I need to know which transcript is most abundant so grabbing all transcripts
myTxiTX <- tximport(files = myquantfiles, type = "salmon", tx2gene = tx2gene, txOut = T)

mytranscripts <- myTxiTX$abundance %>% as.data.frame() %>%
  rownames_to_column(var = "transcript_id")

mytranscripts <- mytranscripts[,-c(5,7)]

#adding the geneID for each transcript
mytranscripts <-  left_join(mytranscripts, tx2gene, by=c("transcript_id" = "transcript_id")) 

mytranscriptsAll <- mytranscripts %>% rowwise() %>% mutate(counts = sum(c_across(`GFP+_ribo_A`:`GFP+_rna_B`))) %>% dplyr::select(transcript_id,gene_id, counts) %>% group_by(gene_id) %>% top_n(1, counts) %>% pull(transcript_id)

#now asking if the geneID is in the list of properly expressed genes and if so, what is the most abundant transcript for that gene. That will now be our transcript of interest for that gene. 
mytranscripts <- mytranscripts %>% filter(gene_id %in% p20TE) %>% rowwise() %>% mutate(counts = sum(c_across(`GFP+_ribo_A`:`GFP+_rna_B`))) %>% dplyr::select(transcript_id,gene_id, counts) %>% group_by(gene_id) %>% top_n(1, counts) %>% pull(transcript_id)

#Now I have the best transcript ID for each, but I don't know the tStart and tStop sites. So now I have to use the gencode v26 GTF file to get that information.

genGTF <- readDNAStringSet(here("accessories", "gencode.v26.pc_transcripts.fa.gz"))

df2 <- data.frame(
  Name = names(genGTF),
  Sequence = as.character(genGTF),
  stringsAsFactors = FALSE
)

# Split the column based on "|"
df2 <- separate(df2, Name, into = paste0("element", 1:10), sep = "\\|", extra = "drop")


TEdata <- df2 %>% filter(element2 %in% p20TE)
TEdata <- TEdata %>% filter(element1 %in% mytranscripts)
TEdata <- TEdata[,-c(3,4,5,6,7,10)]

#throw out any where there is no 5'UTR
TEdata <- TEdata[!grepl("CDS:1-", TEdata$element8),]

#trying to get just the start/stop coordinates
TEdata <- separate(TEdata, element9, into = paste0("element", 3:4), sep = "\\:", extra = "drop")

TEdata <- separate(TEdata, element4, into = paste0("tCodon", 1:2), sep = "\\-", extra = "drop")

TEdata <- TEdata %>%
  mutate(extracted_sequence = mapply(getNAseq, element1, tCodon1, tCodon2))

TEdata$tCodon1 <- as.integer(TEdata$tCodon1)

TEdata <- TEdata %>%
  mutate(Kozak_sequence = mapply(getKozakseq, element1, tCodon1))

write.csv(TEdata, here("NAseqs", "Top20TE_NAseqs.csv"))

```


```{r Ribocode, message=FALSE, warning=FALSE}
Ribocode <- read.table(here("ORFcalling", "Output from ORFcalling Software", "RiboCode_ORFs_result_BcellsAll.txt"), header = TRUE, sep = "\t")

#keep only the Ribocode data that made the filtering process
Ribocode <- Ribocode %>% filter(ORF_ID %in% AllORFs$ORF_ID)
Ribocode <- Ribocode[,c(1:3,11,12)]


#apply the function to every row of Ribocode and add the output as a new column.
Ribocode <- Ribocode %>%
  mutate(extracted_sequence = mapply(getNAseq, transcript_id, ORF_tstart, ORF_tstop))


#apply the function to every row of Ribocode and add the output as a new column.
Ribocode <- Ribocode %>%
  mutate(Kozak_sequence = mapply(getKozakseq, transcript_id, ORF_tstart))

write.csv(Ribocode, here("NAseqs", "Ribocode_NAseqs.csv"))
```



```{r ORFquant, warning=FALSE, message=FALSE}
ORFquant <- read_csv(here("ORFcalling", "Output from ORFcalling Software", "ORFquantExtracted.csv"))

#Need to switch the ORF ID from transcripts to genes for ORFquant
ORFquant <- ORFquant %>% separate(col = "id", into = c("transcript_id", "first", "second"), sep = "_", remove = TRUE) 

 ORFquant <- ORFquant %>% unite(col = "ORF_ID", c("gene", "first", "second"), sep = "_", remove = FALSE)
 
#keep only the Ribocode data that made the filtering process
ORFquant <- ORFquant %>% filter(ORF_ID %in% AllORFs$ORF_ID)

#apply the function to every row of Ribocode and add the output as a new column.
ORFquant <- ORFquant %>%
  mutate(extracted_sequence = mapply(getNAseq, transcript_id, first, second))


ORFquant$first <- as.integer(ORFquant$first)

#apply the function to every row of Ribocode and add the output as a new column.
ORFquant <- ORFquant %>%
  mutate(Kozak_sequence = mapply(getKozakseq, transcript_id, first))


write.csv(ORFquant, here("NAseqs", "ORFquant_NAseqs.csv"))

```

