---
title: "5-Ribowaltz"
author: "Kathryn Walters"
date: "2023-05-18"
output: html_document
---


```{r message=FALSE, warning=FALSE, echo=FALSE}
library(devtools)
install_github("LabTranslationalArchitectomics/riboWaltz", dependencies = TRUE)

library(riboWaltz)
library(GenomicFeatures)
library(here)
library(here)
library(tidyverse)
library(ribosomeProfilingQC)
library(AnnotationDbi)
library(Rsamtools)
library(BSgenome.Hsapiens.UCSC.hg38)
library(ggplot2)
library(ggthemes)
library(GenomicFeatures)
library(ggforce)
```



```{r}
## Summary of RMD contents
#In this file I am using the program riboWaltz to analyze the ribosome profiling data to determine quality. We are looking for periodicity, ~30nt fragment length, and strong enrichment for CDS sequences. The data here is used in Figure 2 and Supplemental Figure 2.  

#1.Loading annotation files and metadata

#2. Creating a metagene plot that shows each replicate separately.

#3. Determining frame density
```


#1.Loading annotation files and metadata


```{r warning=FALSE, message=FALSE}
txdb <- loadDb(here("accessories","txdb.gencode26.sqlite")) #you need your own database file as this is too large for github.

CDS <- prepareCDS(txdb)
genome <- Hsapiens

bamfilename <- Sys.glob(here("data", "bam","*2.bam")) #grabbing bam files. These are not stored on git as too large. Use riboPipe.bsub found in scripts to generate yourself. 

names(bamfilename) <- gsub(x = basename(bamfilename), pattern = "_R2.bam", replacement = "") #cleaning up title. 

```


# 2. Creating a metagene plot that shows each replicate separately.


```{r metaGenePlot, warning=FALSE, message=FALSE}
#here I want to look at the coverage depth for the UTRs and CDS region for each bam file. Then I can take that information and make my own metagene plot where I can see how these compare for each replicate (library 85 and 86)

#getting coverage counts for each region.
cvgs.utr5 <- coverageDepth(RPFs = bamfilename, gtf = txdb, region="utr5")
cvgs.CDS <- coverageDepth(RPFs = bamfilename, gtf = txdb, region="cds")
cvgs.utr3 <- coverageDepth(RPFs = bamfilename, gtf = txdb, region="utr3")

meta_1 <- metaPlot(cvgs.utr5, cvgs.CDS, cvgs.utr3, sample=c(1))
meta_2 <- metaPlot(cvgs.utr5, cvgs.CDS, cvgs.utr3, sample=c(2))

#binding the replicates into one dataframe for each region. 
test <- cbind(as.data.frame(meta_1$UTR5), as.data.frame(meta_2$UTR5))
colnames(test) <- c("Ribo_A", "Ribo_B")

test2 <- cbind(as.data.frame(meta_1$CDS), as.data.frame(meta_2$CDS))
colnames(test2) <- c("Ribo_A", "Ribo_B")

test3 <- cbind(as.data.frame(meta_1$UTR3), as.data.frame(meta_2$UTR3))
colnames(test3) <- c("Ribo_A", "Ribo_B")

#combining all regions into one graph and cleaning column names. 
test4 <- rbind(test,test2,test3)
test4 <- rownames_to_column(test4)
test4$rowname <- as.numeric(test4$rowname)
test4 <- test4[order(test4$rowname),]

#making an average column in case that is needed. 
test4$avg <- rowMeans(test4[,c(2,3)])

#plotting the metaplot. This was not used in final figures. 100 and 600 are the bins that form the cutoffs from the CDS to UTRs. 
ggplot(test4, aes(x=rowname, group = 1)) + geom_line(aes(y=Ribo_A), color = "darkred") +  geom_line(aes(y=Ribo_B), color = "navyblue") + theme_minimal() + geom_vline(xintercept = c(100,600), color = "grey") + xlab("") + ylab("Mean of Coverage") + ggtitle("RP vs IP Metagene Plot - 0hr")

```


# 3. Determining frame density


```{r}
#running routine analysis for library NM2022_0085
myBam <- bamfilename[1]
myBamName <- names(bamfilename[1])
yieldSize <- 1000000
bamfile <- BamFile(myBam, yieldSize = yieldSize)

pSite_best_1 <- estimatePsite(bamfile, CDS, genome)
pc <- getPsiteCoordinates(bamfile, bestpsite = pSite_best_1)
pc.sub_1 <- pc[pc$qwidth %in% c(27,28,29)]
pc.sub_1 <- readsDistribution(pc.sub_1, txdb, las=2)
pc.sub_1 <- assignReadingFrame(pc.sub_1, CDS)
plotDistance2Codon(pc.sub_1)
plotFrameDensity(pc.sub_1)

#running routine analysis for library NM2022_0086
myBam <- bamfilename[2]
myBamName <- names(bamfilename[2])
yieldSize <- 1000000
bamfile <- BamFile(myBam, yieldSize = yieldSize)

pSite_best_5 <- estimatePsite(bamfile, CDS, genome)
pc <- getPsiteCoordinates(bamfile, bestpsite = pSite_best_5)
pc.sub_5 <- pc[pc$qwidth %in% c(27,28,29)]
pc.sub_5 <- readsDistribution(pc.sub_5, txdb, las=2)
pc.sub_5 <- assignReadingFrame(pc.sub_5, CDS)
plotDistance2Codon(pc.sub_5)
plotFrameDensity(pc.sub_5)


#used mergeBams.bsub script to generate a merged bam file. Need to analyze that also. 
bamfilename <- Sys.glob(here("data","bam", "*out.bam"))

myBam <- bamfilename[1]
myBamName <- names(bamfilename[1])
yieldSize <- 1000000
bamfile <- BamFile(myBam, yieldSize = yieldSize)


pSite_best_6 <- estimatePsite(bamfile, CDS, genome)
pc <- getPsiteCoordinates(bamfile, bestpsite = pSite_best_6)
pc.sub_6 <- pc[pc$qwidth %in% c(27,28,29)]
pc.sub_6 <- readsDistribution(pc.sub_6, txdb, las=2)
ggsave(here("plots","Figure2", "SuppFig2B_readDistributionLocation_MergedBams.pdf"), units = "in", width = 6, height = 5, dpi = 320)
pc.sub_6 <- assignReadingFrame(pc.sub_6, CDS)
plotDistance2Codon(pc.sub_6)
plotFrameDensity(pc.sub_6)
p_save <- plotFrameDensity(pc.sub_6)

#saving bar graph of frame density.
p_save <- p_save %>% as.data.frame()

ggplot(data=p_save, aes(x = Var1, y = Freq)) +
  geom_bar(stat="identity") + theme_minimal() + ylab("Relative Read Density (%)") + xlab("Frame")

ggsave(here("plots","Figure2", "Fig2B_FrameDensity_MergedBams.pdf"),units = "in", width = 4, height = 3, dpi = 320)

```

