---
title: "5-Ribowaltz"
author: "Kathryn Walters"
date: "2023-05-18"
output: html_document
---


```{r message=FALSE, warning=FALSE, echo=FALSE}
library(devtools)
install_github("LabTranslationalArchitectomics/riboWaltz", dependencies = TRUE)

library(riboWaltz)
library(GenomicFeatures)
library(here)
library(here)
library(tidyverse)
library(ribosomeProfilingQC)
library(AnnotationDbi)
library(Rsamtools)
library(BSgenome.Hsapiens.UCSC.hg38)
library(ggplot2)
library(ggthemes)
# library(Cairo)
library(GenomicFeatures)
library(ggforce)


txdb <- loadDb(here("accessories","txdb.gencode26.sqlite"))
CDS <- prepareCDS(txdb)
genome <- Hsapiens

bamfilename <- Sys.glob(here("data", "bam","*.bam"))
names(bamfilename) <- gsub(x = basename(bamfilename), pattern = "_R2.bam", replacement = "")


cvgs.utr5 <- coverageDepth(RPFs = bamfilename, gtf = txdb, region="utr5")
cvgs.CDS <- coverageDepth(RPFs = bamfilename, gtf = txdb, region="cds")
cvgs.utr3 <- coverageDepth(RPFs = bamfilename, gtf = txdb, region="utr3")
meta_1 <- metaPlot(cvgs.utr5, cvgs.CDS, cvgs.utr3, sample=c(1))
meta_2 <- metaPlot(cvgs.utr5, cvgs.CDS, cvgs.utr3, sample=c(2))


test <- cbind(as.data.frame(meta_1$UTR5), as.data.frame(meta_2$UTR5))
colnames(test) <- c("Ribo_A", "Ribo_B")

test2 <- cbind(as.data.frame(meta_1$CDS), as.data.frame(meta_2$CDS))
colnames(test2) <- c("Ribo_A", "Ribo_B")

test3 <- cbind(as.data.frame(meta_1$UTR3), as.data.frame(meta_2$UTR3))
colnames(test3) <- c("Ribo_A", "Ribo_B")

test4 <- rbind(test,test2,test3)
test4 <- rownames_to_column(test4)
test4$rowname <- as.numeric(test4$rowname)
test4 <- test4[order(test4$rowname),]
#write_csv(test4, here("output", "CoverageData_0hr_all.csv"))

#test4 <- read_csv(here("output", "CoverageData_0hr_all.csv"), col_names = TRUE)
test4$avg <- rowMeans(test4[,c(2,3)])

ggplot(test4, aes(x=rowname, group = 1)) + geom_line(aes(y=Ribo_A), color = "darkred") +  geom_line(aes(y=Ribo_B), color = "navyblue") + theme_minimal() + geom_vline(xintercept = c(100,600), color = "grey") + xlab("") + ylab("Mean of Coverage") + ggtitle("RP vs IP Metagene Plot - 0hr")

p_meta_0hr_zoom <- ggplot(test4, aes(x=rowname, group = 1)) + geom_line(aes(y=Ribo_A), color = "darkred") +  geom_line(aes(y=Ribo_B), color = "navyblue") + theme_minimal() + geom_vline(xintercept = c(100), color = "grey") + xlab("") + ylab("Mean of Coverage") + xlim(95,130)


```


```{r}
myBam <- bamfilename[1]
myBamName <- names(bamfilename[1])
yieldSize <- 1000000
bamfile <- BamFile(myBam, yieldSize = yieldSize)

pSite_best_1 <- estimatePsite(bamfile, CDS, genome)
pc <- getPsiteCoordinates(bamfile, bestpsite = pSite_best_1)
pc.sub_1 <- pc[pc$qwidth %in% c(27,28,29)]
pc.sub_1 <- readsDistribution(pc.sub_1, txdb, las=2)
pc.sub_1 <- assignReadingFrame(pc.sub_1, CDS)
plotDistance2Codon(pc.sub_1)
plotFrameDensity(pc.sub_1)


myBam <- bamfilename[2]
myBamName <- names(bamfilename[2])
yieldSize <- 1000000
bamfile <- BamFile(myBam, yieldSize = yieldSize)

pSite_best_5 <- estimatePsite(bamfile, CDS, genome)
pc <- getPsiteCoordinates(bamfile, bestpsite = pSite_best_5)
pc.sub_5 <- pc[pc$qwidth %in% c(27,28,29)]
pc.sub_5 <- readsDistribution(pc.sub_5, txdb, las=2)
pc.sub_5 <- assignReadingFrame(pc.sub_5, CDS)
plotDistance2Codon(pc.sub_5)
plotFrameDensity(pc.sub_5)




bamfilename <- Sys.glob(here("data","*.bam"))
names(bamfilename) <- gsub(x = basename(bamfilename), pattern = "_R2.bam", replacement = "")


myBam <- bamfilename[1]
myBamName <- names(bamfilename[1])
yieldSize <- 1000000
bamfile <- BamFile(myBam, yieldSize = yieldSize)


pSite_best_6 <- estimatePsite(bamfile, CDS, genome)
pc <- getPsiteCoordinates(bamfile, bestpsite = pSite_best_6)
pc.sub_6 <- pc[pc$qwidth %in% c(27,28,29)]
pc.sub_6 <- readsDistribution(pc.sub_6, txdb, las=2)
ggsave(here("plots", "readDistributionLocation_MergedBams.pdf"), units = "in", width = 6, height = 5, dpi = 320)
pc.sub_6 <- assignReadingFrame(pc.sub_6, CDS)
plotDistance2Codon(pc.sub_6)
plotFrameDensity(pc.sub_6)

p_save <- p_save %>% as.data.frame()

ggplot(data=p_save, aes(x = Var1, y = Freq)) +
  geom_bar(stat="identity") + theme_minimal() + ylab("Relative Read Density (%)") + xlab("Frame")

ggsave(here("plots", "FrameDensity_MergedBams.jpg"),units = "in", width = 4, height = 3, dpi = 320)

plotDistance2Codon(pc.sub_1)
plotDistance2Codon(pc.sub_5)
plotDistance2Codon(pc.sub_6)
plotDistance2Codon(pc.sub_t1)

#ribotricer uses: 
#Betacell_mergedBamsSorted.sortedByCoord.out.bam

#ribocode uses: 
#NM2022_0085_R2Aligned.toTranscriptome.out.bam




bamfilename <- Sys.glob(here("data", "bam_test", "*.bam"))
names(bamfilename) <- gsub(x = basename(bamfilename), pattern = "_R2.bam", replacement = "")

myBam <- bamfilename[1]
myBamName <- names(bamfilename[1])
yieldSize <- 1000000
bamfile <- BamFile(myBam, yieldSize = yieldSize)


pSite_best_t1 <- estimatePsite(bamfile, CDS, genome)
pc <- getPsiteCoordinates(bamfile, bestpsite = pSite_best_t1)
pc.sub_t1 <- pc[pc$qwidth %in% c(27,28,29)]
pc.sub_t1 <- readsDistribution(pc.sub_t1, txdb, las=2)
pc.sub_t1 <- assignReadingFrame(pc.sub_t1, CDS)

plotFrameDensity(pc.sub_t1)



```

