---
title: "6-Translational Efficiency"
author: "Kathryn Walters"
date: "2023-05-11"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r message=FALSE, warning=FALSE}
library(tidyverse)
library(here)
library(msigdbr)
library(clusterProfiler)
library(ggplot2)
library(ggpubr)
library(cowplot)
library(ggVennDiagram)
library(scales)
library(ggthemes)
library(preprocessCore)
library(dplyr)
library(writexl)
```


```{r}
## Summary of RMD contents
#In this file I am calculating the translational efficiency (Ribosomal Footprint reads/ Total RNA reads). I'm then using these values to characterize the cell lines used and find that they align with what would be expected for insulin producing beta cells. The data here is used in Figure 2.

#1. Loading annotation files and metadata

#2. Calculating TE 

#3. GSEA for enriched TE genes

#4. Function/Code to Visualize TE for Specific Genes
```


#1. Loading annotation files and metadata


```{r load data, message=FALSE, warning=FALSE, echo=FALSE}
#grabbing and creating necessary files
geneInfo <- read_csv(here::here("accessories","gencode.v26.primary.info.csv.zip"), col_names = F, show_col_types = FALSE) 
geneInfo <- geneInfo[,c(1,3,4)] %>% distinct()
colnames(geneInfo) <- c("gene_id","biotype","symbol")

annotation <- geneInfo %>% 
  dplyr::select(gene_id, symbol, biotype) %>% # exclude transcripts
  unique() %>% # remove duplicates
  as.data.frame() # convert into dataframe

#gene symbols only
gene_symbol <- geneInfo %>% dplyr::select(gene_id, symbol) %>% unique()

#What datafiles do I need? TPM normalized to calculate TE
allCountsCPM <- read_csv(here::here("output", "allCounts_CPMnorm.csv"))
allCountsTPM <- read_csv(here::here("output", "allCounts_TPMnorm.csv"))
```


# 2.Calculating and plotting TE 


```{r calculate TE, message=FALSE, warning=FALSE, echo=FALSE}
#TE is the ribosomal footprints/rna. 

#take the average of the replicates and then divide and take log2
TEdata <- allCountsTPM[,c(1,2:4,6)]
TEdata$meanRibo <- rowMeans(TEdata[ , c(2,3)])
TEdata$meanRNA <- rowMeans(TEdata[ , c(4,5)] )

TE_norm <- TEdata[,c(1,6,7)] %>% column_to_rownames("gene_id")

TE_norm <- as.data.frame(normalize.quantiles(as.matrix(TE_norm))) #normalizing the data
TE_norm <- cbind(TEdata[,1], TE_norm)
colnames(TE_norm) <- c("gene_id", "NormMeanRibo", "NormMeanRNA")

TE_norm$TE <- TE_norm$NormMeanRibo/TE_norm$NormMeanRNA

TEdata <- left_join(TEdata, TE_norm)
TEdata$log10_TE <- (log10(TEdata$TE))

#removing any infinite values
TEdata <- TEdata[!is.infinite(abs(rowSums(TEdata[,10]))), ]

TEdata <- left_join(TEdata, geneInfo) #%>% filter(biotype == "protein_coding")

#writexl::write_xlsx(TEdata,  here("output", "TPM_TEdata.xlsx")) #saving the normalized TE calculations

#want to be able to compare the TE of all genes to those of specific biotypes. The expectation is that lncRNAs will have a lower TE since they are not traditionally translated. 
TEdata2 <- TEdata %>%
  mutate(BiotypeII = case_when(biotype == 'lincRNA' ~ as.character(biotype), biotype == 'protein_coding' ~ as.character(biotype), .default = 'other'))

TEdata2$BiotypeII <- factor(TEdata2$BiotypeII, levels = c("other", "lincRNA","protein_coding"))

TEviolinDensity <- ggviolin(data = TEdata2 %>% filter(BiotypeII != "other"), x = "BiotypeII", y = "TE", trim = F, add=c("mean")) +
  ylab("Normalized TE") +
  xlab("") +
  scale_y_continuous(trans="log10",
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x)), limits = c(0.001,500)) +
  annotation_logticks(sides = "l") +
  theme_minimal()

TEviolinDensity

TEdata2 %>% filter(BiotypeII != "other") %>% group_by(BiotypeII) %>% summarise(mean(TE))

#add statistics - parametric two sample t-test

statsData <- TEdata2 %>% filter(BiotypeII != "other")

t.test(statsData %>% filter(BiotypeII == "protein_coding") %>% dplyr::select(log10_TE), statsData %>% filter(BiotypeII == "lincRNA") %>% dplyr::select(log10_TE), alternative = "greater", var.equal = FALSE)

#ggsave(plot = TEviolinDensity, filename = here::here("plots", "Figure2", "Fig2C_TE_violinDensity.pdf"), device = "pdf", units = "in", width = 5, height = 4, dpi = 320)

```


# 3. GSEA for enriched TE genes


```{r exploratory GSEA, message=FALSE, warning=FALSE, echo=FALSE}
h_gene_sets = msigdbr(species = "human") #getting hallmark gene sets.

TEdata <- TEdata[!is.infinite(abs(rowSums(TEdata[,2:9]))), ]


TEdata_proteinCoding <- TEdata %>% filter(biotype == "protein_coding" & log10_TE > -10) 
DEgeneList <- TEdata_proteinCoding %>% pull(log10_TE) #grabbing TE change for ALL genes (above threshold) in this experiment. 
names(DEgeneList) <- TEdata_proteinCoding %>% pull(symbol)
DEgeneList <- sort(DEgeneList, decreasing = TRUE) #need to be sorted by decreasing value. 

msigdbr_t2g <- h_gene_sets %>% dplyr::distinct(gs_name, gene_symbol) %>% as.data.frame()

#running GSEA test
set.seed(42)
edo1 <- GSEA(geneList = DEgeneList, pAdjustMethod = "fdr", 
               pvalueCutoff = 1,
               minGSSize = 2,
               maxGSSize = 4000,
               TERM2GENE = msigdbr_t2g)
tmp1 <- edo1@result
tmp1

HallmarkGenes <- tmp1 %>% filter(p.adjust < 0.05)



CP_gene_sets = msigdbr(species = "human", category = "C2") #getting CP gene sets.

msigdbr_t2g_CP <- CP_gene_sets %>% dplyr::distinct(gs_name, gene_symbol) %>% as.data.frame()

#running GSEA test
set.seed(42)
edo2 <- GSEA(geneList = DEgeneList, pAdjustMethod = "fdr", 
               pvalueCutoff = 1,
               minGSSize = 5,
               maxGSSize = 4000,
               TERM2GENE = msigdbr_t2g_CP)
tmp2 <- edo2@result

CP_Genes <- tmp2 %>% filter(p.adjust < 0.01)
KEGG_Genes <- CP_Genes %>% filter(str_detect(ID, "^KEGG"))


KEGG_Genes <- separate(data = KEGG_Genes, col = ID, sep = "KEGG_", into = c("hallmark", "ID"))
plotGSEAexplore <- KEGG_Genes %>%
  arrange(NES) %>%    # First sort by val. This sort the dataframe but NOT the factor levels
  mutate(ID=factor(ID, levels=ID)) %>%   # This trick updates the factor levels
 ggplot( aes(x=ID, y=NES, fill = p.adjust)) +
    geom_bar(stat="identity", alpha=1, width=.6) +
    coord_flip() +
    labs(title = "GSEA Hallmark Enrichment") + xlab(NULL) +
    theme_minimal() + scale_fill_gradient(low = "black", high = "#E69F00")

plotGSEAexplore

ggsave(plot = plotGSEAexplore, filename = here::here("plots", "Figure2", "Fig2D_TE_GSEAPlotexplore.pdf"), device = "pdf", units = "in", width = 8, height = 5, dpi = 320)

```


# 4. Function/Code to Visualize TE for Specific Genes


```{r specific TE genes, message=FALSE, warning=FALSE}
#This was not used for the publication but is useful for those wanting to visualize specific genes. 

#creating a function
plotgeneTPM <- function(tpms, goi) {

mygene <- annotation %>% dplyr::filter(symbol==goi)
tpmsGene <- tpms %>% as.data.frame() %>% dplyr::filter(gene_id == mygene$gene)

tpmsGene <- pivot_longer(data = tpmsGene, cols = 2:7) %>% separate(col = "name", into = c("type", "exp", "rep"), sep = "_")
tpmsGene <- tpmsGene %>% filter(type != "GFP-")

ggplot(tpmsGene, aes(x=exp, y=value)) +
    geom_dotplot(binaxis='y', stackdir='center', dotsize = 1.5) +
    ylab("CPM") +
    xlab("") +
    theme_minimal() + ggtitle(mygene$symbol)
}

#Now creating plots
p_GCG <- plotgeneTPM(allCountsCPM, "GCG")

p_INS <- plotgeneTPM(allCountsCPM, "INS")
p_SOX9 <- plotgeneTPM(allCountsCPM, "SOX9")

p_PAX4 <- plotgeneTPM(allCountsCPM, "PAX4")




p_PAX6 <- plotgeneTPM(allCountsCPM, "PAX6")

plotgeneTPM(allCountsCPM, "PAK3")
plotgeneTPM(allCountsCPM, "FOXA2")
plotgeneTPM(allCountsCPM, "GCK")


p_GenePlotsBcells <- plot_grid(p_INS  + theme(legend.position="none"),
                       p_PAX6 + theme(legend.position="none"),
                        p_GCG  + theme(legend.position="none"),
                                          nrow = 1)

p_GenePlotsBcells

```

