---
title: "7-ORFcalling Dataset Collapsing"
author: "Kathryn Walters"
date: "2023-05-11"
output: html_document
---
```{r libraries, warning=FALSE, message=FALSE, echo=FALSE}
library(Biostrings)
library(rlist)
library(ggVennDiagram)
library(readr)
library(here)
library(dplyr)
```


```{r}
## Summary of RMD contents
#In this file my goal is to combine the ORFs called by Ribocode and ORFquant into one dataset. I will use gffcompare to compare them and filter from there. gffCompare was run on command line and the files are available in the accessories folder. The data here is used in downstream RMD files and analysis.

#1. Loading data

#2. Using gffCompare to combine datasets
  #Created the database final_collapsedORFsList.csv which is the final output after using gffCompare to collapse everything. 
  #Also created the database GeneCollapsedORFs_AllPrograms.csv in case I want a list of what each program called. 
```


# 1. Loading data


```{r load data, message=FALSE, warning=FALSE, echo=FALSE}
ribocode <- read_csv(here("output", "Expression_Filtered_ORFcalling", "RiboCodeFiltered.csv"))

ORFquant <- read_csv(here("output", "Expression_Filtered_ORFcalling", "ORFquant_Filtered.csv"))

```


# 2. Using gffCompare to combine datasets


```{r gtf combined, warning=FALSE, message=FALSE}
#this one has ORFquant as the reference file.
gffCompare_ALL <- read_tsv(here("accessories", "gffCompare", "tryRef.Ribocode_renamed_final_GNchanged2.gtf.tmap"))

#this one has ribocode as the reference file.
gffCompare_ALL2 <- read_tsv(here("accessories", "gffCompare", "tryRef_OQ.ORFquant_CDS_final_GNchanged2.gtf.tmap"))
IDswitch <- read_csv(here("accessories", "ORFquant_IDswitchResource.csv")) #need to switch the IDs from transcript to gene level to match Ribocode. 


#making lists of things that are marked the same (class code "=") by gffCompare. 
gffCompare1 <- gffCompare_ALL[,c(3,4)] %>% filter(class_code == "=") %>% dplyr::select(qry_gene_id) %>% unique()

gffCompare2 <- gffCompare_ALL2[,c(1,3)] %>% filter(class_code == "=") %>% dplyr::select(ref_gene_id) %>% unique()

#comparing if these geneIDs are the same based on which file was used as reference. 
differences <- setdiff(gffCompare2$ref_gene_id, gffCompare1$qry_gene_id)
differences

differences1 <- setdiff(gffCompare1$qry_gene_id, gffCompare2$ref_gene_id)
differences1

#So these 6 are ones that are sometimes marked = but not in both datasets. Therefore, we are going to keep them as separate. We are going to change these class codes in the necessary datasets. 
ORFquantProblems <- c("ENST00000255380.8_780_2549", "ENST00000617528.1_27_404", "ENST00000409556.5_366_4328", "ENST00000281317.9_1217_2464", "ENST00000281830.3_611_994", "ENST00000619474.4_991_1425")

gffCompare_ALL2 <- gffCompare_ALL2 %>% dplyr::mutate(class_codeII = case_when(qry_gene_id %in% ORFquantProblems ~ "c", .default = class_code))


#Now let's try to make one dataframe..
ribocode <- left_join(ribocode, gffCompare_ALL, by = c("ORF_ID" = "qry_gene_id"))

ribocode <- ribocode %>% mutate(keepColumn = case_when(class_code == "=" ~ "Both", .default = "Ribocode"))



gffCompare_ALL2 <- left_join(gffCompare_ALL2, IDswitch, by = c("qry_gene_id" = "ORF_ID_transcript"))

ORFquant <- left_join(ORFquant, gffCompare_ALL2, by = c("ORF_ID" = "ORF_ID_gene"))

ORFquant <- ORFquant %>% mutate(keepColumn = case_when(class_codeII == "=" ~ "Both", .default = "ORFquant"))


OQ1 <- ORFquant[,c(1,2,3,4,17,18)]
RC1 <- ribocode[,c(1,2,3,4,7,16)]

colnames(OQ1) <- c("ORF_ID", "AAseq", "ORF_type", "gene_id", "class_code", "keepColumn")
colnames(RC1) <- c("ORF_ID", "AAseq", "ORF_type", "gene_id", "class_code", "keepColumn")

final_nuORFs <- rbind(RC1, OQ1 %>% filter(keepColumn == "ORFquant"))

#looking at what classifications we have left. Many are the same ORF called, but some unique. This is expected. 
final_nuORFs %>% pull(class_code) %>% table()

#outputting file with collapsed ORFs. 
write_csv(final_nuORFs, here("output", "final_collapsedORFsList.csv"))


#just in case I want a full list of anything ever called by either software, I will create that also. 
ribocode <- read_csv(here("output", "Expression_Filtered_ORFcalling", "RiboCodeFiltered.csv"))
ORFquant <- read_csv(here("output", "Expression_Filtered_ORFcalling", "ORFquant_Filtered.csv"))

ribocode$program <- "Ribocode"
ORFquant$program <- "ORFquant"

colnames(ORFquant) <- c("ORF_ID", "AAseq", "ORF_type", "gene_id", "program")

ALLORFs <- rbind(ribocode, ORFquant)

write_csv(ALLORFs, here("output", "GeneCollapsedORFs_AllPrograms.csv"))

```

