---
title: "9-nuORF Collapsing"
author: "Kathryn Walters"
date: "2023-05-17"
output: html_document
editor_options: 
  chunk_output_type: console
---


```{r libraries, warning=FALSE, message=FALSE, echo=FALSE}
library(Biostrings)
library(rlist)
library(ggVennDiagram)
library(ggpubr)
library(scales)
library(readr)
library(here)
library(dplyr)
library(tidyr)
library(ggplot2)
library(ggthemes)
library(readxl)
library(ggokabeito)
library(rstatix)
library(tibble)
library(preprocessCore)
```


```{r}
## Summary of RMD contents
#In this file I am characterizing the nuORFs that were identified. For all of the listed categories below, I look at the trends faceted by program (in case there was program dependency), but then finally just as a whole group of ORFs as program didn't seem to make a difference for any category. The data here is used in Figure 3 and Supplementary Figure 3. 

#1. Loading Data

#2. Loading proteomics data

#3. Separating nuORFs from annoORFs

#4. Identifing nuORFs that are called by both programs

#5. nuORF Proteomic Support

#6. nuORF Length

#7. Characterizing the "novel" category

#8. RNA expression

#9. Ribosome Occupancy - Translational Efficiency

#10. Proteomic Intensity
```


# 1. Loading Data


```{r load data, message=FALSE, warning=FALSE, echo=FALSE}
AllORFs <- read_csv(here::here("output", "GeneCollapsedORFs_AllPrograms.csv"))

collapseORFs <- read_csv(here::here("output", "final_collapsedORFsList.csv"))

geneInfo <- read_csv(here::here("accessories","gencode.v26.primary.info.csv.zip"), col_names = F, show_col_types = FALSE) 
colnames(geneInfo) <- c("gene_id","transcript_id","biotype","symbol")

tx2gene <- geneInfo[,c(2,1)]
```


# 2. Loading proteomics data


```{r proteomics load, warning=FALSE, message=FALSE, echo=FALSE}
#getting RiboCode list of ORFs found in sBC proteomics
proteomicsRC <- read_tsv(here::here("proteomics", "RiboCode", "sBC_nORF_Ribocode", "combined_protein.tsv")) %>% filter(`Combined Unique Spectral Count` > 1) %>% as.data.frame()
proteomicsRC <- proteomicsRC[,c(2,26,27)]
colnames(proteomicsRC) <- c("ORF_ID", "Unique_Intensity_1", "Unique_Intensity_2")
proteomicsRC$Average_Intensity <- (proteomicsRC$Unique_Intensity_1 + proteomicsRC$Unique_Intensity_2)/2


#getting ORFquant list of ORFs found in proteomics
proteomicsOQ <- read_tsv(here::here("proteomics", "ORFQuant", "sBC_nORF_ORFquant", "combined_protein.tsv")) %>% filter(`Combined Unique Spectral Count` > 1) %>% as.data.frame()
proteomicsOQ <- proteomicsOQ[,c(2,26,27)]
colnames(proteomicsOQ) <- c("ORF_ID", "Unique_Intensity_1", "Unique_Intensity_2")
proteomicsOQ$Average_Intensity <- (proteomicsOQ$Unique_Intensity_1 + proteomicsOQ$Unique_Intensity_2)/2
```


# 3. Separating nuORFs from annoORFs


```{r per software, message=FALSE, warning=FALSE, echo=FALSE}
#grabbing only NON-annotated genes
NOTannoORFs <- AllORFs %>% filter(ORF_type != "annotated" & ORF_type != "other")

RC_NotAnno <- NOTannoORFs %>% filter(program == "Ribocode")
OQ_NotAnno <- NOTannoORFs %>% filter(program == "ORFquant")


orfTable <- AllORFs %>% group_by(program) %>% select(ORF_type) %>% table() %>% as.data.frame()

orfTable$ORF_type <- factor(orfTable$ORF_type, order = TRUE, levels =c('other', 'dORF', "uORF", "novel", "annotated"))
levels(orfTable$ORF_type)

#making plots
p_nuORFs_percent <- ggplot(orfTable %>% filter(ORF_type != "annotated"), aes(fill=ORF_type, y=Freq, x=program)) + 
    geom_bar(position="fill", stat="identity") + theme_minimal() + scale_fill_colorblind() + ylab("Percent") + xlab("")

p_nuORFs_percent

#ggsave(plot = p_nuORFs_percent, filename = here("plots", "Figure3", "Fig3C_nuORFs_percent.pdf"), width = 4, height = 3)

```


# 4. Identifing nuORFs that are called by both programs 


```{r overlap seqs, message=FALSE, warning=FALSE, echo=FALSE}
#identifying nuORFs that have the EXACT SAME sequence. This is a very conservative approach because they could be nested or not quite calling the same start/stop but essentially be the same. 
set1 <- as.matrix(RC_NotAnno)[,2]
set3 <- as.matrix(OQ_NotAnno)[,2]

listALL <- list(RiboCode=set1, ORFquant = set3)

p_nuORFoverlap <- ggVennDiagram(listALL, edge_color = "black", label = "count", label_alpha = 0) + theme(legend.position = "none")

p_nuORFoverlap

#ggsave(plot = p_nuORFoverlap, filename = here("plots", "Figure3", "Fig3B_nuORFoverlap_venndiagram.pdf"), width = 4, height = 3)

```


#Needed??

```{r message=FALSE, warning=FALSE}
Top20TE <- read_csv(here::here("NAseqs", "Top20TE_NAseqs.csv"))

#ORF_ID, AA_seq, ORF_type, gene_id, program, protein - AllORFs
#column 4, 9, "Top20TE", 3, "None", "Yes"

Top20TE <- Top20TE[,c(4,9,3)]
colnames(Top20TE) <- c("ORF_ID", "AAseq", "gene_id")

Top20TE$ORF_type <- "Top20TE"
Top20TE$program <- "None"
Top20TE$protein <- "Yes"

Top20TE <- Top20TE %>% select("ORF_ID", "AAseq", "ORF_type", "gene_id", "program", "protein")

colnames(Top20TE)
colnames(AllORFs)
AllORFs <- rbind(AllORFs, Top20TE)
```


# 5. nuORF Proteomic Support


```{r proteomic support, message=FALSE, warning=FALSE, echo=FALSE}
# First we will look at proteomic support for each program separately.

#combine the list of identified genes from both programs
allproteomics <- rbind(proteomicsRC, proteomicsOQ)

AllORFs <- AllORFs %>% dplyr::mutate(protein = case_when(ORF_ID %in% allproteomics$ORF_ID ~ 'Yes', !(ORF_ID %in% allproteomics$ORF_ID) ~ 'No'))
AllORFs$protein %>% table()

tableProteomicSupport <- AllORFs %>% group_by(ORF_type, program) %>% select(protein) %>% table() %>% as.data.frame()

tableProteomicSupport$ORF_type <- factor(tableProteomicSupport$ORF_type, order = TRUE, levels =c('annotated', 'novel', 'uORF', "dORF", "other"))
levels(tableProteomicSupport$ORF_type)

p_nuORF_proteoSupport_percent <- ggplot(tableProteomicSupport %>% filter(ORF_type != "annotated" & ORF_type != "other"), aes(fill=protein, y=Freq, x=program)) + 
    geom_bar(position="fill", stat="identity") + theme_minimal() + scale_fill_colorblind() + ylab("Percent") + xlab("") +
  facet_grid(~ ORF_type) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

p_nuORF_proteoSupport_percent


#Now let's collapse the programs and just look in general. 

tableProteomicSupport_collapsed <- AllORFs %>% group_by(ORF_type) %>% select(protein) %>% table() %>% as.data.frame()

tableProteomicSupport_collapsed$ORF_type <- factor(tableProteomicSupport_collapsed$ORF_type, order = TRUE, levels =c('novel', 'uORF', "dORF", "other"))
levels(tableProteomicSupport_collapsed$ORF_type)

p_nuORF_proteoSupport_percent_collapsed <- ggplot(tableProteomicSupport_collapsed %>% filter(ORF_type != "annotated" & ORF_type != "other"), aes(fill=protein, y=Freq, x=ORF_type)) + 
    geom_bar(position="fill", stat="identity") + theme_minimal() + scale_fill_colorblind() + ylab("Percent") + xlab("") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

p_nuORF_proteoSupport_percent_collapsed

#ggsave(plot = p_nuORF_proteoSupport_percent_collapsed, filename = here("plots", "Figure3", "Fig3G_nuORFProteinSupport_percent_collapsed.pdf"), width = 5, height = 3)

p_nuORF_proteoSupport_count <- ggplot(tableProteomicSupport %>% filter(ORF_type != "annotated" & ORF_type != "other"), aes(fill=protein, y=Freq, x=program)) + 
    geom_bar(position="stack", stat="identity") + theme_minimal() + scale_fill_colorblind() + ylab("Count") + xlab("") +
  facet_grid(~ ORF_type) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

p_nuORF_proteoSupport_count

#ggsave(plot = p_nuORF_proteoSupport_count, filename = here("plots", "Figure3", "Supp3D_nuORFProteinSupport_counts.pdf"), width = 5, height = 3)
```


# 6. nuORF Length


```{r nuORF length, message=FALSE, warning=FALSE}

AllORFs$nchar <- nchar(AllORFs$AAseq)

AllORFs$ORF_type <- factor(AllORFs$ORF_type, order = TRUE, levels =c('annotated', 'novel', 'uORF', "dORF", "other"))

AllORFs$protein <- factor(AllORFs$protein, order = TRUE, levels =c('Yes', 'No'))

colors <- c("#969696", "#009E73", "#56B4E9", "#E69F00", "#CC79A7")

AllORFs$charLog <- log10(AllORFs$nchar)

p_nuORF_lengthAA_anno2 <- ggviolin(AllORFs %>% filter(ORF_type != "other" & ORF_type != "Top20TE"), y="nchar", x="program", fill = "ORF_type", trim = F, add=c("mean"))+ theme_minimal() + scale_fill_manual(values = colors) + ylab("ORF Length - AA") + xlab("") + facet_grid( ~ ORF_type) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + scale_y_continuous(trans="log10",
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x))) +
  annotation_logticks(sides = "l") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

p_nuORF_lengthAA_anno2

p_nuORF_lengthAA_anno <- ggviolin(AllORFs %>% filter(ORF_type != "other"), y="nchar", x="ORF_type", fill = "ORF_type", trim = F, add=c("mean"))+ theme_minimal() + scale_fill_manual(values = colors) + ylab("ORF Length - AA") + xlab("") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + scale_y_continuous(trans="log10",
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x))) +
  annotation_logticks(sides = "l") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

p_nuORF_lengthAA_anno


#ggsave(plot = p_nuORF_lengthAA_anno, filename = here("plots", "Figure3", "Fig3D_nuORF_length_anno.pdf"), width = 5, height = 4)


#Wilcoxon rank sum test since data is non-parametric. Tells difference between ORF types overall.

kruskal.test(nchar ~ ORF_type, data = AllORFs)

data2 <- AllORFs
pairwise.wilcox.test(data2$nchar, data2$ORF_type,
                 p.adjust.method = "BH")

#Wilcoxon rank sum test since data is non-parametric. Tells difference between ORF types overall per program.

data2 <- AllORFs %>% filter(ORF_type == "annotated")
pairwise.wilcox.test(data2$nchar, data2$program,
                 p.adjust.method = "BH")
data2 %>%
  group_by(program) %>%
  summarise_at(vars(nchar), list(name = mean))


data2 <- AllORFs %>% filter(ORF_type == "novel")
pairwise.wilcox.test(data2$nchar, data2$program,
                 p.adjust.method = "BH")

data2 %>%
  group_by(program) %>%
  summarise_at(vars(nchar), list(name = mean))

data2 <- AllORFs %>% filter(ORF_type == "uORF")
pairwise.wilcox.test(data2$nchar, data2$program,
                 p.adjust.method = "BH")

data2 %>%
  group_by(program) %>%
  summarise_at(vars(nchar), list(name = mean))

data2 <- AllORFs %>% filter(ORF_type == "dORF")
pairwise.wilcox.test(data2$nchar, data2$program,
                 p.adjust.method = "BH")

data2 %>%
  group_by(program) %>%
  summarise_at(vars(nchar), list(name = mean))
```


# 7. Characterizing the "novel" category


```{r characterize novel category, message=FALSE, warning=FALSE}

geneInfo <- geneInfo[,c(1,3,4)] %>% unique()

geneType <- left_join(AllORFs, geneInfo)

geneType <- geneType %>% filter(ORF_type == "novel")

geneType <- geneType %>% dplyr::mutate(biotypeII = case_when(
  biotype == "processed_pseudogene" ~ 'pseudogene', 
  biotype == "transcribed_processed_pseudogene" ~ 'pseudogene', 
  biotype == "unprocessed_pseudogene" ~ 'pseudogene', 
  biotype == "transcribed_unitary_pseudogene" ~ 'pseudogene', 
  biotype == "transcribed_unprocessed_pseudogene" ~ 'pseudogene', 
  biotype == "3prime_overlapping_ncRNA" ~ 'lncRNA', 
  biotype == "sense_intronic" ~ 'lncRNA', 
  biotype == "processed_transcript" ~ 'lncRNA', 
  biotype == "TEC" ~ 'lncRNA', 
  biotype == "antisense" ~ 'lncRNA', 
  biotype == "lincRNA" ~ 'lncRNA', 
  biotype == "sense_overlapping" ~ 'lncRNA',
  .default = biotype))

geneType %>% select(biotypeII) %>% table()

geneType$biotypeII <- factor(geneType$biotypeII, order = TRUE, levels =c("protein_coding", 'lncRNA', 'pseudogene'))

geneTypeNovel <- ggplot(geneType, aes(x=biotypeII)) + 
    geom_bar(position="stack") + theme_minimal() + ylab("Count") + xlab("") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

geneTypeNovel

#ggsave(plot = geneTypeNovel, filename = here("plots", "Figure3", "SuppFig3A_nuORF_geneTypeNovel.pdf"), width = 5, height = 4)
```


# 8. RNA expression


```{r RNA expression, message=FALSE, warning=FALSE, echo=FALSE}

#Now I want to get the expression of each gene associated with each nuORF category and see if some nuORF categories are coming from genes with high or low expression. Going to use the GFP+ RNA-seq average of replicates for this number. 

#Start with program separate information. 

TPMnorm <- read_csv(here::here("output", "allCounts_TPMnorm.csv"))

TPMnormGFP <- TPMnorm[,c(1,4,6)]
TPMnormGFP$GFP_avg <- (TPMnormGFP$`GFP+_rna_A`+ TPMnormGFP$`GFP+_rna_B`)/2
TPMnormGFP <- TPMnormGFP[,c(1,4)]


AllORFs <- left_join(AllORFs, TPMnormGFP)

AllORFs_violin <- AllORFs %>% select(gene_id, ORF_type, GFP_avg, program, protein) %>% unique()

p_violinNUORFrnaExpress_program <- ggviolin(data = AllORFs_violin %>% filter(ORF_type != "other"), x = "program", y = "GFP_avg", trim = F, add=c("mean"), fill = "ORF_type") +
  ylab("log10(rnaTPM)") +
  xlab("") +
  scale_y_continuous(trans="log10",
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x)), limits = c(.05,1000)) +facet_grid(~ ORF_type) +
  annotation_logticks(sides = "l") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + scale_fill_manual(values = colors) +
  theme_minimal()

p_violinNUORFrnaExpress_program

#ggsave(plot = p_violinNUORFrnaExpress_program, filename = here("plots", "nuORF_rnaExpressProgram_violin.pdf"), width = 6, height = 3)


#Wilcoxon rank sum test since data is non-parametric. Tells difference between ORF types overall per program.

data2 <- AllORFs %>% filter(ORF_type == "annotated")
pairwise.wilcox.test(data2$GFP_avg, data2$program,
                 p.adjust.method = "BH")
data2 %>% filter(GFP_avg > 0) %>% 
  group_by(program) %>%
  summarise_at(vars(GFP_avg), list(name = mean))

data2 <- AllORFs %>% filter(ORF_type == "novel")
pairwise.wilcox.test(data2$GFP_avg, data2$program,
                 p.adjust.method = "BH")

data2 %>% filter(GFP_avg > 0) %>% 
  group_by(program) %>%
  summarise_at(vars(GFP_avg), list(name = mean))

data2 <- AllORFs %>% filter(ORF_type == "uORF")
pairwise.wilcox.test(data2$GFP_avg, data2$program,
                 p.adjust.method = "BH")

data2 %>% filter(GFP_avg > 0) %>% 
  group_by(program) %>%
  summarise_at(vars(GFP_avg), list(name = mean))

data2 <- AllORFs %>% filter(ORF_type == "dORF")
pairwise.wilcox.test(data2$GFP_avg, data2$program,
                 p.adjust.method = "BH")

data2 %>% filter(GFP_avg > 0) %>% 
  group_by(program) %>%
  summarise_at(vars(GFP_avg), list(name = mean))


#collapsing the programs
AllORFs_violin2 <- AllORFs %>% select(gene_id, ORF_type, GFP_avg, protein) %>% unique()

AllORFs_violin2 %>% pull(ORF_type) %>% table()

p_violinNUORFrnaExpress <- ggviolin(data = AllORFs_violin2 %>% filter(ORF_type != "other"), x = "ORF_type", y = "GFP_avg",
trim = F, add=c("mean"), fill = "ORF_type") +
  ylab("log10(rnaTPM)") +
  xlab("") +
  scale_y_continuous(trans="log10",
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x)), limits = c(.05,1500)) +
  annotation_logticks(sides = "l") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + scale_fill_manual(values = colors) +
  theme_minimal()

p_violinNUORFrnaExpress

#ggsave(plot = p_violinNUORFrnaExpress, filename = here("plots","Figure3", "Supp3B_nuORF_rnaExpress_violin.pdf"), width = 5, height = 3)


#Wilcoxon rank sum test since data is non-parametric. Tells difference between ORF types overall.

kruskal.test(GFP_avg ~ ORF_type, data = AllORFs)

data2 <- AllORFs
pairwise.wilcox.test(data2$GFP_avg, data2$ORF_type,
                 p.adjust.method = "BH")

```


# 9. Ribosome Occupancy - Translational Efficiency


```{r ribosome occupancy, message=FALSE, warning=FALSE, echo=FALSE}
#Now I want to get the ribosome counts of each gene and see if some categories are coming from genes with high or low translation Going to use the average of replicates for this number. 

#take the average of the replicates and then divide and take log2
TEdata <- TPMnorm[,c(1,2:4,6)]
TEdata$meanRibo <- rowMeans(TEdata[ , c(2,3)])
TEdata$meanRNA <- rowMeans(TEdata[ , c(4,5)] )

TE_norm <- TEdata[,c(1,6,7)] %>% column_to_rownames("gene_id")

TE_norm <- as.data.frame(normalize.quantiles(as.matrix(TE_norm)))
TE_norm <- cbind(TEdata[,1], TE_norm)
colnames(TE_norm) <- c("gene_id", "NormMeanRibo", "NormMeanRNA")

TE_norm$TE <- TE_norm$NormMeanRibo/TE_norm$NormMeanRNA

TE_norm$log10_TE <- log10(TE_norm$TE)

#TE_norm <- left_join(TE_norm, geneInfo) #%>% filter(biotype == "protein_coding")

#TEdata2 <- TEdata[,c(1,8)]

AllORFs <- left_join(AllORFs, TE_norm)

AllORFs_violin_ribo <- AllORFs %>% select(gene_id, ORF_type, TE, program, protein) %>% unique() %>% na.omit()

AllORFs_violin_ribo$ORF_type <- factor(AllORFs_violin_ribo$ORF_type, order = TRUE, levels =c('annotated', 'novel', 'uORF', "dORF", "other"))

colors <- c("#969696", "#009E73", "#56B4E9", "#E69F00", "#CC79A7", "#551A8B")


p_violinNUORF_TE <- ggviolin(data = AllORFs_violin_ribo %>% filter(ORF_type != "other"), x = "ORF_type", y = "TE", fill = "ORF_type",
trim = F, add=c("mean")) + geom_hline( yintercept = 1.01) + 
  ylab("Normalized TE") +
  xlab("") +
  scale_y_continuous(trans="log10",
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x)), limits = c(.05, 50)) +
  theme_minimal() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + annotation_logticks(sides = "l") + scale_fill_manual(values = colors)

p_violinNUORF_TE

#ggsave(plot = p_violinNUORF_TE, filename = here::here("plots", "Figure3", "Fig3E_nuORF_TE_violin.pdf"), width = 5, height = 3)


#Wilcoxon rank sum test since data is non-parametric.

kruskal.test(TE ~ ORF_type, data = AllORFs_violin_ribo)

data2 <- AllORFs_violin_ribo
pairwise.wilcox.test(data2$TE, data2$ORF_type,
                 p.adjust.method = "BH")

AllORFs_violin_ribo %>% group_by(ORF_type) %>% dplyr::summarise(mean(log10(TE)))

#separating by program too, in case there is a difference. 

p_violinNUORF_TE_program <- ggviolin(data = AllORFs_violin_ribo %>% filter(program != "None" & ORF_type != "other"), x = "program", y = "TE", fill = "ORF_type",
trim = F, add=c("mean")) +
  ylab("Normalized TE") +
  xlab("") + facet_wrap(~ORF_type, nrow = 1) +
  scale_y_continuous(trans="log10",
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x)), limits = c(.01, 100)) +
  theme_minimal() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + annotation_logticks(sides = "l") + scale_fill_manual(values = colors) 

p_violinNUORF_TE_program


data2 <- AllORFs %>% filter(ORF_type == "annotated")
pairwise.wilcox.test(data2$TE, data2$program,
                 p.adjust.method = "BH")

data2 %>% filter(TE > 0) %>% 
  group_by(program) %>%
  summarise_at(vars(TE), list(name = mean))

data2 <- AllORFs %>% filter(ORF_type == "novel")
pairwise.wilcox.test(data2$TE, data2$program,
                 p.adjust.method = "BH")

data2 %>% filter(TE > 0) %>% 
  group_by(program) %>%
  summarise_at(vars(TE), list(name = mean))

data2 <- AllORFs %>% filter(ORF_type == "uORF")
pairwise.wilcox.test(data2$TE, data2$program,
                 p.adjust.method = "BH")

data2 %>% filter(TE > 0) %>% 
  group_by(program) %>%
  summarise_at(vars(TE), list(name = mean))

data2 <- AllORFs %>% filter(ORF_type == "dORF")
pairwise.wilcox.test(data2$TE, data2$program,
                 p.adjust.method = "BH")

data2 %>% filter(TE > 0) %>% 
  group_by(program) %>%
  summarise_at(vars(TE), list(name = mean))

```


# 10. Proteomic Intensity


```{r warning=FALSE, message=FALSE}
#want to get a feel for the level of protein intensity/expression for each of the categories. Maybe the nuORFs were barely expressed. Plus, some categories, such as uORFs are more typically regulatory and thus should have an overall lower protein expression level. 

AllORFs2 <- left_join(AllORFs, allproteomics) %>% na.omit() %>% filter(Average_Intensity > 0)

AllORFs2$ORF_type <- factor(AllORFs2$ORF_type, order = TRUE, levels =c('annotated', 'novel', 'uORF', "dORF", "other"))

AllORFs2$logValue <- log10(AllORFs2$Average_Intensity)

p_nuORF_proteomicIntensity <- ggviolin(data = AllORFs2 %>% filter(ORF_type != "other" & ORF_type != "Top20TE"), x = "ORF_type", y = "Average_Intensity", fill = "ORF_type",
trim = F, add=c("mean")) +
  ylab("Total Intensity") +
  xlab("") +
  scale_y_continuous(trans="log10",
    breaks = trans_breaks("log10", function(x) 10^x),
    labels = trans_format("log10", math_format(10^.x))) +
  theme_minimal() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + annotation_logticks(sides = "l")+ scale_fill_manual(values = colors) 

p_nuORF_proteomicIntensity

#ggsave(plot = p_nuORF_proteomicIntensity, filename = here("plots", "Figure3", "Fig3H_nuORF_proteomicIntensity.pdf"), width = 6, height = 3)

kruskal.test(Average_Intensity ~ ORF_type, data = AllORFs2)

data2 <- AllORFs2
pairwise.wilcox.test(data2$Average_Intensity, data2$ORF_type,
                 p.adjust.method = "BH")



data2 <- AllORFs2 %>% filter(ORF_type == "annotated")
pairwise.wilcox.test(data2$Average_Intensity, data2$program,
                 p.adjust.method = "BH")

data2 <- AllORFs2 %>% filter(ORF_type == "novel")
pairwise.wilcox.test(data2$Average_Intensity, data2$program,
                 p.adjust.method = "BH")

data2 <- AllORFs2 %>% filter(ORF_type == "uORF")
pairwise.wilcox.test(data2$Average_Intensity, data2$program,
                 p.adjust.method = "BH")

AllORFs2 %>% filter(ORF_type == "uORF") %>%
  group_by(program) %>%
  summarise_at(vars(Average_Intensity), list(name = mean))

data2 <- AllORFs2 %>% filter(ORF_type == "dORF")
pairwise.wilcox.test(data2$Average_Intensity, data2$program,
                 p.adjust.method = "BH")

nuORFS_ALLdata <- read_csv(here("output", "AllORFs_nu_ALLDATA.csv"))

nuORFS_ALLdata <- left_join(nuORFS_ALLdata, AllORFs2, by = "ORF_ID")

```

